services/api/src/utils/logging.js

```diff
@@ -1,22 +1,133 @@
-function timeStr() {
-  return new Date().toString();
+const pino = require('pino');
+const config = require('@bedrockio/config');
+
+const tracer = require('./tracer');
+const startTime = Symbol('startTime');
+
+function getGlobalContext() {
+  const currentSpan = tracer.getCurrentSpan();
+  if (!currentSpan) return null;
+  return currentSpan.context();
 }
 
-const logger = {
-  debug(...args) {
-    console.info(...[timeStr(), '-', 'DEBUG', '-'].concat(args));
-  },
-  info(...args) {
-    console.info(...[timeStr(), '-', 'INFO', '-'].concat(args));
+const parentLogger = pino({
+  messageKey: 'message',
+  formatters: {
+    level(_, level) {
+      if (level === 20) {
+        return { severity: 'debug' };
+      }
+      if (level === 30) {
+        return { severity: 'info' };
+      }
+      if (level === 40) {
+        return { severity: 'warning' };
+      }
+      if (level === 50) {
+        return { severity: 'error' };
+      }
+      if (level >= 60) {
+        return { severity: 'critical' };
+      }
+      return { severity: 'default' };
+    },
   },
-  warn(...args) {
-    console.info(...[timeStr(), '-', 'WARN', '-'].concat(args));
-  },
-  error(...args) {
-    console.info(...[timeStr(), '-', 'ERROR', '-'].concat(args));
+  base: null,
+  level: config.get('LOG_LEVEL'),
+  timestamp: pino.stdTimeFunctions.isoTime,
+  prettyPrint: process.env.NODE_ENV === 'dev' ? { messageKey: 'message' } : false,
+});
+
+function formatCurrentTrace({ traceId, spanId }) {
+  return {
+    'logging.googleapis.com/spanId': spanId,
+    'logging.googleapis.com/trace': traceId,
+    'logging.googleapis.com/trace_sampled': true,
+  };
+}
+
+const formatters = {
+  gcloud: function ({ request, response, latency }) {
+    // https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry
+    return Object.assign({
+      message: `${request.method} ${request.url} ${response.getHeader('content-length')} - ${latency}ms`,
+      httpRequest: {
+        requestMethod: request.method.toUpperCase(),
+        requestUrl: request.url,
+        requestSize: request.headers['content-length'],
+        status: response.statusCode,
+        userAgent: request.headers['user-agent'],
+        referer: request.headers['referer'],
+        remoteIp: request.headers['x-forwarded-for'],
+        latency: `${Math.floor(latency / 1e3)}.${Math.floor(latency % 1e3)}s`,
+        protocol: request.headers['x-forwarded-proto'],
+        responseSize: response.getHeader && response.getHeader('content-length'),
```

Comment:
are you expecting `getHeader` to sometimes be null here? you're looking at the headers in a few different ways here including `content-length` itself... wondering what's expected input

Reply:
i can remove that, not required. 

the requestSize and responseSize is naturally quite different also in terms of api. 
e.g. response https://nodejs.org/dist/latest-v14.x/docs/api/http.html#http_class_http_serverresponse 
vs https://nodejs.org/dist/latest-v14.x/docs/api/http.html#http_class_http_incomingmessage 

the incoming message doesn't expose a getHeader method but just object of headers. 
https://nodejs.org/dist/latest-v14.x/docs/api/http.html#http_message_headers

Reply:
oh I see I hadn't realized it was two different objects... looks good 👍